### PR TITLE
(Towards #1374 ) Childless/Leaf directive

### PR DESCRIPTION
**Pull request by LonelyCat124**
This contains the initial childless/leaf directive implementation for #1374 . It is based upon the non-master branch, so is dependent on #1368 to be merged.

The idea is to handle directives which don't wrap around code, e.g.
```
!$omp taskwait
```
by removing the subtree which will never be used for codegen.

@arporter also thinks this can be applied to certain ACC directives, e.g. `ACCEnterDataDirective`. I don't see where the base implementation of that class uses its child schedule. Similarly for its child, `GOACCEnterDataDirective`. However, upon application to the `ACCEnterDataDirective` the build is faililng for a reason I don't understand. @sergisiso does this type of node seem reasonable to you?

### CODE REVIEW

**Review by sergisiso — CHANGES_REQUESTED** [review]
@AidanChalk I made a first look at the implementation and added a few inline comments. You will see the main comment is that I think we don't need all the intermediate node_str and dag_name methods as these should be abstract classes and probably don't need this functionality*.
(* I actually think that this 2 methods have to be slightly changed so they are only needed on the node super-class and avoid all the boilerplate code, but I will submit another issue for this) 

Other than the in-line comment I am still worried about OMPDirective not being a mixin, not because the code duplication which I think there isn't*, but because the `node.walk(OMPDirective)` will not pick up every OpenMP directive.
There are 2 solutions:
1) Is to change the name of OMPDirective and ACCDirective to OMPRegionDirective and ACCRegionDirective, so it is obvious in the walk method call that it will be missing things. But this will require to update all the `node.walk(OMPDirective)`-style code (some of it in scripts outside the codebase :/)
2) Use Mixin classes like:
```
class OMPDirective():
       _PREFIX = OMP

class OMPTaskwaitDirective(StandaloneDirective, OMPDirective):
     ... Same code ...

class OMPSerialDirective(RegionDirective, OMPDirective)
     ... Same code ...
```
and the same for the ACCDirectives tree.

\* I think the OMPDirective._get_reductions_list and the ACCDirective.validate_globals_constraints can both be moved to the generic RegionDirective as they are not specific of OMP or ACC. Also the node_str and dag_name as I mentioned earlier. So there is no code other than the prefix attribute for the Mixin which solve a few issues.

**Comment by LonelyCat124** [comment]
Ok - that makes sense I think. I'll try switching to Mixin classes and adjust the rest of the changes.

**Comment by sergisiso** [comment]
Note that I created #1383, so if removing some intermediate dag_name and node_str creates some problems in this PR just leave them and it will be tacked in that next issue.

File: src/psyclone/psyir/backend/c.py [hunk 1]
```diff
@@ -476,3 +477,20 @@ def ompdirective_node(self, node):
         self._depth -= 1
         result_list.append("}\n")
         return "".join(result_list)
+
+    def ompstandalonedirective_node(self, node):
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/backend/c.py:481]: I think the ompdirective_node already existed, I'll change this to RegionDirective and StandaloneDirective I think, it makes more sense and should work for ACC.

**Comment by sergisiso** [comment]
You could put a `import pdb; pdb.set_trace()` in the visitor and it will open the debugger then when running the tests. Then you can check what is in the prefix attribute or do `bt` to get the backtrace.

Btw, have you removed the `_PREFIX = ""` in the generic directive, I think attributes on the first class resolution path have priority.

**Comment by LonelyCat124** [comment]
> Btw, have you removed the `_PREFIX = ""` in the generic directive?, I think attributes on the first class resolution path have priority.

This was the issue thanks!

**Comment by sergisiso** [comment]
I see no reason to prevent the generation of C OpenACC pragmas if these are correct even if its a use case not fully tested yet.

File: src/psyclone/psyir/backend/c.py [hunk 2]
```diff
@@ -476,3 +477,20 @@ def ompdirective_node(self, node):
...
+        '''This method is called when an OMPStandaloneDirective instance is
+        found in the PSyIR tree. It returns the opening and closing directives,
+        and the statements in between as a string (depending on the language).
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/backend/c.py:484]: Done

File: src/psyclone/psyir/backend/c.py [hunk 3]
```diff
@@ -476,3 +477,20 @@ def ompdirective_node(self, node):
...
+
+        :param node: a StandaloneDirective PSyIR node.
+        :type node: :py:class:`psyclone.psyir.nodes.StandaloneDirective`
+
+        :returns: the C code as a string.
+        :rtype: str
+
+        '''
+        # pylint: disable=no-self-use
+        # Note that {{ is replaced with a single { in the format call
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/backend/c.py:494]: Done

File: src/psyclone/psyir/backend/fortran.py [hunk 4]
```diff
@@ -1325,6 +1326,22 @@ def directive_node(self, node):
             result_list.append("{0}!${1}\n".format(self._nindent, end_string))
         return "".join(result_list)
 
+    def standalonedirective_node(self, node):
+        '''This method is called when a StandaloneDirective instance is found
+        in the PSyIR tree. It returns the opening and closing directives, and
+        the statements in between as a string (depending on the language).
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/backend/fortran.py:1332]: Done

File: src/psyclone/psyir/backend/c.py [hunk 5]
```diff
@@ -476,3 +477,20 @@ def ompdirective_node(self, node):
...
+        result_list = ["#pragma {0}\n".format(node.begin_string())]
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/backend/c.py:495]: Done

File: src/psyclone/psyir/nodes/directive.py [hunk 6]
```diff
@@ -35,11 +35,14 @@
 #         I. Kavcic,    Met Office
 #         C.M. Maynard, Met Office / University of Reading
 #         J. Henrichs, Bureau of Meteorology
+# Modified A. B. G. Chalk, STFC Daresbury Lab.
 # -----------------------------------------------------------------------------
 
 ''' This module contains the Directive node implementation.'''
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/directive.py:41]: Done

File: src/psyclone/psyir/nodes/directive.py [hunk 7]
```diff
@@ -50,11 +53,23 @@
...
 
+@six.add_metaclass(abc.ABCMeta)
 class Directive(Statement):
     '''
-    Base class for all Directive statements.
+    Abstract base class for all Directive statements.
 
-    All classes that generate Directive statements (e.g. OpenMP,
+    '''
+    # The prefix to use when constructing this directive in Fortran
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/directive.py:62]: I've changed this comment (as I had to remove `_PREFIX`. Its a bit weird now though, not sure if `_PREFIX=None` would work with the new mixins or not.

File: src/psyclone/psyir/nodes/directive.py [hunk 8]
```diff
@@ -50,11 +53,23 @@
...
+    # (e.g. "OMP"). Must be set by sub-class.
+    _PREFIX = ""
+    _text_name = "Directive"
+
+
+class RegionDirective(Directive):
+    '''
+    Based class for all Region Directive statements.
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/directive.py:70]: Done

File: src/psyclone/psyir/nodes/directive.py [hunk 9]
```diff
@@ -246,5 +259,55 @@ def _add_region(self, start_text, end_text=None, data_movement=None):
...
+
+    '''
+    # Textual description of the node.
+    _children_valid_format = None
+    _text_name = "StandaloneDirective"
+    _colour = "green"
+
+    def __init__(self, ast=None, children=None, parent=None):
+        super(StandaloneDirective, self).__init__(ast, children=children,
+                                                  parent=parent)
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/directive.py:286]: Done

File: src/psyclone/psyir/nodes/directive.py [hunk 10]
```diff
@@ -246,5 +259,55 @@ def _add_region(self, start_text, end_text=None, data_movement=None):
...
+        :returns: the name to use in the DAG for this node.
+        :rtype: str
+        '''
+        _, position = self._find_position(self.ancestor(Routine))
+        return "standalone_directive_" + str(position)
+
+
 # For automatic API documentation generation
-__all__ = ["Directive"]
+__all__ = ["Directive", "RegionDirective", "StandaloneDirective"]
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/directive.py:313]: None of them can be put at the file top level (have checked) due to circular dependencies I think.

I think the error is changed to run like you've said now.

File: src/psyclone/psyir/nodes/omp_directives.py [hunk 11]
```diff
@@ -112,25 +113,42 @@ def _get_reductions_list(self, reduction_type):
...
-            raise GenerationError(
-                "OMPTaskwaitDirective was provided children. This"
-                " directive does not support children. See #1374"
-                " for more information.")
-        super(OMPTaskwaitDirective, self).__init__(children=children,
-                                                   parent=parent)
+    Class representing an OpenMP TASKWAIT directive in the PSyIR.
+
+    '''
+    _children_valid_format = None
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/omp_directives.py:151]: Done

File: src/psyclone/psyir/nodes/omp_directives.py [hunk 12]
```diff
@@ -112,25 +113,42 @@ def _get_reductions_list(self, reduction_type):
...
     '''
-    Class representing an OpenMP TASKWAIT directive in the PSyIR.
+    Base class for all OpenMP-related standalone directives
 
-    :param list children: None. See #1374 for details.
-    :param parent: The Node in the AST that has this directive as a child.
-    :type parent: :py:class:`psyclone.psyir.nodes.Node`
+    '''
+    _PREFIX = "OMP"
+    _children_valid_format = None
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/omp_directives.py:122]: Done

File: src/psyclone/psyir/nodes/directive.py [hunk 13]
```diff
@@ -50,11 +53,23 @@
...
     '''
-    Base class for all Directive statements.
+    Abstract base class for all Directive statements.
 
-    All classes that generate Directive statements (e.g. OpenMP,
+    '''
+    # The prefix to use when constructing this directive in Fortran
+    # (e.g. "OMP"). Must be set by sub-class.
+    _PREFIX = ""
+    _text_name = "Directive"
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/directive.py:65]: I think I've done this now.

**Comment by LonelyCat124** [comment]
Should I make the OMPRegion and OMPStandalone directives abstract classes again?

File: src/psyclone/psyir/nodes/omp_directives.py [hunk 14]
```diff
@@ -60,7 +61,7 @@
 OMP_OPERATOR_MAPPING = {AccessType.SUM: "+"}
 
 
-class OMPDirective(Directive):
+class OMPDirective(RegionDirective):
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/omp_directives.py:64]: I've removed them, I believe so.

File: src/psyclone/psyir/nodes/omp_directives.py [hunk 15]
```diff
@@ -112,25 +113,42 @@ def _get_reductions_list(self, reduction_type):
...
+        to generate coloured output in a terminal that supports it.
+
+        :param bool colour: whether or not to include colour control codes.
 
-    :raises GenerationError: if this OMPTaskwait is provided children. \
-                             See #1374 for details.
+        :returns: description of this node, possibly coloured.
+        :rtype: str
+        '''
+        return self.coloured_name(colour) + "[OMPStandalone]"
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/omp_directives.py:143]: Done

**Comment by LonelyCat124** [comment]
Pylint is not super happy about the mixing classes as they have no public methods.

The `RegionDirective` expects the `_PREFIX` member to exist in its definition. The issue with this being in the mixin class is that it searches (as Sergi explained to me) the parent class up the Directives tree and finds that before using the Mixin's definition. I'm not sure what happens if I define _PREFIX as None here but I assume the same issue would arise. 
I think this is only a linting issue and the code works, but it seems a bit nasty. @sergisiso if I declared things as `class OMPStandaloneDirective(OMPDirective, StandaloneDirective)` would that solve the issue? Is that an appropriate fix if so?

**Comment by LonelyCat124** [comment]
Fingers crossed the tests all passs now :)

**Comment by sergisiso** [comment]
@LonelyCat124 Changing the order of the classes is a good idea, give it a shoot to see if it improves things. If that allows adding the generic _PREFIX back it will be nicer from a OO point of view. But if not, no worries about the pylint issue if the reason, also you can ignore the "not enough methods" one.

**Comment by LonelyCat124** [comment]
The order swap seemed to work (at least for my local test runs) so I've done that and added `_PREFIX` back to the Directive class.

**Review by sergisiso — CHANGES_REQUESTED** [review]
@LonelyCat124 I like how the implementation turned out with the mixin and swaping the parent class order. This is almost ready, I just have some remaining issues when testing with "termcolor" and there are some minor clean ups to do (see comments about both things inline).

File: src/psyclone/psyir/backend/c.py [hunk 16]
```diff
@@ -456,23 +457,43 @@ def loop_node(self, node):
                "{0}}}\n".format(self._nindent, variable_name,
                                 start, stop, step, body)
 
-    def ompdirective_node(self, node):
-        '''This method is called when an OMPDirective instance is found in
+    def regiondirective_node(self, node):
+        '''This method is called when an REgionDirective instance is found in
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/backend/c.py:461]: Done

File: src/psyclone/psyir/backend/c.py [hunk 17]
```diff
@@ -456,23 +457,43 @@ def loop_node(self, node):
...
+        :param node: a RegionDirective PSyIR node.
+        :type node: :py:class:`psyclone.psyir.nodes.RegionDirective`
 
         :returns: the C code as a string.
         :rtype: str
 
         '''
         # Note that {{ is replaced with a single { in the format call
-        result_list = ["#pragma {0}\n{{\n".format(node.begin_string())]
+        print(type(node))
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/backend/c.py:473]: Yep, oops!

File: src/psyclone/psyir/nodes/acc_directives.py [hunk 18]
```diff
@@ -45,39 +46,35 @@
...
 from psyclone.psyir.nodes.structure_reference import StructureReference
 from psyclone.psyir.symbols import DataSymbol
 from psyclone.core import AccessType, VariablesAccessInfo
 
 
-class ACCDirective(Directive):
-    ''' Base class for all OpenACC directive statements. '''
+@six.add_metaclass(abc.ABCMeta)
+class ACCDirective():
+    ''' Base mixin class for all OpenACC directive statements. '''
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/acc_directives.py:60]: Done, though I changed class to class's (apparently the correct possessive of a singular class but looks weird to me anyway).

File: src/psyclone/psyir/nodes/omp_directives.py [hunk 19]
```diff
@@ -60,34 +61,21 @@
 OMP_OPERATOR_MAPPING = {AccessType.SUM: "+"}
 
 
-class OMPDirective(Directive):
+@six.add_metaclass(abc.ABCMeta)
+class OMPDirective():
     '''
-    Base class for all OpenMP-related directives
+    Base mixin class for all OpenMP-related directives.
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/omp_directives.py:67]: Done

File: src/psyclone/tests/psyir/nodes/directive_test.py [hunk 20]
```diff
@@ -96,23 +98,35 @@ def test_directive_backward_dependence():
...
     # Additional children
     with pytest.raises(GenerationError) as excinfo:
         directive.addchild(schedule)
-    assert ("Item 'Schedule' can't be child 1 of 'Directive'. The valid format"
-            " is: 'Schedule'." in str(excinfo.value))
+    assert ("Item 'Schedule' can't be child 1 of 'RegionDirective'. The valid "
+            "format is: 'Schedule'." in str(excinfo.value))
+
+
+def test_childlessdirective_children_validation():
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/tests/psyir/nodes/directive_test.py:123]: Done

File: src/psyclone/tests/psyir/nodes/acc_directives_test.py [hunk 21]
```diff
@@ -381,12 +387,11 @@ def test_accupdatedirective_init():
 
 
 def test_accupdatedirective_begin_and_end_strings():
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/tests/psyir/nodes/acc_directives_test.py:389]: Done

File: src/psyclone/psyir/nodes/directive.py [hunk 22]
```diff
@@ -67,18 +85,15 @@ class Directive(Statement):
...
 
     '''
-    # The prefix to use when constructing this directive in Fortran
-    # (e.g. "OMP"). Must be set by sub-class.
-    _PREFIX = ""
     # Textual description of the node.
     _children_valid_format = "Schedule"
-    _text_name = "Directive"
-    _colour = "green"
+    _text_name = "RegionDirective"
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply src/psyclone/psyir/nodes/directive.py:90]: Done. Some tests are now flagged with TODO associated with #1383 as error messages no longer specify the class spawning the error message.

**Comment by LonelyCat124** [comment]
Ok, I think I fixed all the issues and merged the latest master.

**Review by sergisiso — APPROVED** [review]
@LonelyCat124 All comments addressed. Tests pass also with compilation, coverage of diff is complete, pycodestyle and pylint pass without relevant new issues. New classes/methods are well documented in docstrings, reference guide builds well and no additional documentation is needed. No TODOs remaining for this issues. This PR is ready to merge.

**Pull request merged**